src/rest-server/src/models/v2/job.js

```diff
@@ -0,0 +1,297 @@
+  // hdfs operations
+  const hdfs = new HDFS(launcherConfig.webhdfsUri);
+  // async mkdir on hdfs
+  const mkdir = async (path, user, mode) => {
+    const options = {'user.name': user, 'permission': mode};
+    return new Promise((resolve) => {
+      hdfs.createFolder(path, options, (err, result) => {
+        if (err) {
+          throw err;
+        } else {
+          resolve(result);
+        }
+      });
+    });
+  };
+  // async upload file on hdfs
+  const upload = async (path, data, user, mode) => {
+    const options = {'user.name': user, 'permission': mode, 'overwrite': 'true'};
+    return new Promise((resolve) => {
+      hdfs.createFile(path, data, options, (err, result) => {
+        if (err) {
+          throw err;
+        } else {
+          resolve(result);
+        }
+      });
+    });
+  };
+
+  // generate framework description
+  const frameworkDescription = generateFrameworkDescription(frameworkName, userName, config);
+
+  // prepare scripts on hdfs
+  const pathPrefix = `/Container/${userName}/${frameworkName}`;
+  await mkdir('/Container', 'root', '777');
+  for (let path of ['log', 'tmp']) {
+    await mkdir(`${pathPrefix}/${path}`, userName, '755');
+  }
+  for (let taskRole of Object.keys(config.taskRoles)) {
+    await upload(`${pathPrefix}/YarnContainerScripts/${taskRole}.sh`,
+      generateYarnContainerScript(frameworkName, userName, config, frameworkDescription, taskRole), userName, '644');
+    await upload(`${pathPrefix}/DockerContainerScripts/${taskRole}.sh`,
+      generateDockerContainerScript(frameworkName, userName, config, taskRole), userName, '644');
+  }
+  // upload framework description file to hdfs
+  await upload(`${pathPrefix}/${launcherConfig.frameworkDescriptionFilename}`,
+    JSON.stringify(frameworkDescription, null, 2), userName, '644');
+  // upload original config file to hdfs
+  await upload(`${pathPrefix}/${launcherConfig.jobConfigFileName}`.replace(/json$/, 'yaml'),
+    rawConfig, userName, '644');
+
+  // generate ssh key
+  if (process.platform.toUpperCase() === 'LINUX') {
+    const keys = await generateSSHKeys(frameworkName);
+    if (keys) {
+      for (let keyname of Object.keys(keys)) {
+        await upload(`${pathPrefix}/ssh/keyFiles/${keyname}`, keys[keyname], userName, '755');
+      }
+    }
+  }
+
+  // return framework description
+  return frameworkDescription;
+};
+
+
+class Job {
+  constructor(jobName, userName, ) {
+    this.jobName = jobName;
+    this.userName = userName;
+    this.frameworkName = this.userName ? `${this.userName}~${this.jobName}` : this.jobName;
+  }
+}
+
+Job.prototype.put = async function(config, rawConfig) {
+  // check user vc
+  const virtualCluster = ('defaults' in config && config.defaults.virtualCluster != null) ?
+    config.defaults.virtualCluster : 'default';
+  await new Promise((resolve) => {
+    userModel.checkUserVc(this.userName, virtualCluster, (err, result) => {
+      if (err) {
+        throw err;
```

Comment:
another `throw` issue, I'm not found all of them so please do not rely on where I comment.

Reply:
updated